build_runner/lib/src/asset_graph/optional_output_tracker.dart

```diff
@@ -0,0 +1,38 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:build/build.dart';
+
+import '../generate/phase.dart';
+import 'graph.dart';
+import 'node.dart';
+
+class OptionalOutputTracker {
+  final _checkedOutputs = <AssetId, bool>{};
+  final AssetGraph _assetGraph;
+  final List<BuildPhase> _buildPhases;
+
+  OptionalOutputTracker(this._assetGraph, this._buildPhases);
+
+  bool isRequired(AssetId output, [Set<AssetId> currentlyChecking]) {
+    currentlyChecking ??= new Set<AssetId>();
+    if (currentlyChecking.contains(output)) return false;
+    currentlyChecking.add(output);
+
+    final node = _assetGraph.get(output);
+    if (node is! GeneratedAssetNode) return true;
+    final generatedNode = node as GeneratedAssetNode;
+    final phase = _buildPhases[generatedNode.phaseNumber];
+    if (!phase.isOptional) return true;
+    return _checkedOutputs.putIfAbsent(
+        output,
+        () =>
+            generatedNode.outputs
+                .any((o) => isRequired(o, currentlyChecking)) ||
+            _assetGraph
```

Comment:
I don't quite get this logic - why do we care about other outputs in the phase?

Reply:
Oh - I think this is trying to handle the additional outputs of the same action?

Right now I think it is checking all outputs of the same phase which is a lot more than we actually want I believe. We could add an additional `where` after `outputsForPhase` that checks if the primary input is the same?

Reply:
https://github.com/dart-lang/build/pull/1034/commits/6dda418fa6879656c73cbcf7cc67da6414037edc

https://github.com/dart-lang/build/pull/1034#issuecomment-367117339

Reply:
Ah I see, yeah that makes sense. Will switch to only those with the same primary input

Reply:
We still need to either do the `expectedOutputs` trick I did there or add a `where` after the `outputsToPhase` though to retain the old logic right?

Reply:
yeah added a where clause to check the primary inputs

Reply:
Lol I really wish github updated comments without having to refresh lol